audit/audit.go

```diff
@@ -0,0 +1,179 @@
+	User            string        `avro:"user" json:"user,omitempty"`
+	AttemptedAction string        `avro:"attempted_action" json:"attempted_action,omitempty"`
+	ActionResult    string        `avro:"action_result" json:"action_result,omitempty"`
+	Params          common.Params `avro:"params" json:"params,omitempty"`
+}
+
+type avroMarshaller func(s interface{}) ([]byte, error)
+
+// OutboundProducer defines a producer for sending outbound audit events to a kafka topic
+type OutboundProducer interface {
+	Output() chan []byte
+}
+
+//AuditorService defines the behaviour of an auditor
+type AuditorService interface {
+	Record(ctx context.Context, action string, result string, params common.Params) error
+}
+
+//Auditor provides functions for interception HTTP requests and populating the context with a base audit event and
+// recording audit events
+type Auditor struct {
+	service       string
+	marshalToAvro avroMarshaller
+	producer      OutboundProducer
+}
+
+// NopAuditor is an no op implementation of the AuditorService.
+type NopAuditor struct{}
+
+//Record is a no op implementation of Auditor.Record
+func (a *NopAuditor) Record(ctx context.Context, attemptedAction string, actionResult string, params common.Params) error {
+	return nil
+}
+
+//New creates a new Auditor with the namespace, producer and token provided.
+func New(producer OutboundProducer, namespace string) *Auditor {
+	log.Debug("auditing enabled for service", log.Data{"service": namespace})
+	return &Auditor{
+		producer:      producer,
+		service:       namespace,
+		marshalToAvro: EventSchema.Marshal,
+	}
+}
+
+//Record captures the provided action, result and parameters and an audit event. Common fields - time, user, service
+// are added automatically. An error is returned if there is a problem recording the event it is up to the caller to
+// decide what do with the error in these cases.
+func (a *Auditor) Record(ctx context.Context, attemptedAction string, actionResult string, params common.Params) error {
+	//NOTE: for now we are only auditing user actions - this may be subject to change
+	user := identity.User(ctx)
+	if user == "" {
+		log.Debug("not user attempted action: skipping audit event", nil)
+		return nil
+	}
+
+	if attemptedAction == "" {
+		return NewAuditError("attemptedAction required but was empty", "", actionResult, params)
+	}
+	if actionResult == "" {
+		return NewAuditError("actionResult required but was empty", attemptedAction, "", params)
+	}
+
+	e := Event{
+		Service:         a.service,
+		User:            user,
+		AttemptedAction: attemptedAction,
+		ActionResult:    actionResult,
+		Created:         time.Now().String(),
+		Params:          params,
+	}
+
+	e.RequestID = requestID.Get(ctx)
+
+	avroBytes, err := a.marshalToAvro(e)
+	if err != nil {
+		log.Error(err, nil)
+		return NewAuditError("error marshalling event to arvo", attemptedAction, actionResult, params)
+	}
+
+	log.Info("logging audit message", log.Data{"auditEvent": e})
```

Comment:
Is there value in this log message being more explicit about this happening before we've sent the kafka message? just thinking when debugging it might help to be clear whether this log message means 'successful audit all the things' or 'weve created something in an audit structure, now we're going to try actually sending it'

Reply:
TBH I only put this log message here so I could see that was something was being "audited" while I was developing it. I will remove it.